File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/205-isomorphic-strings.py
def is_isomorphic(s: str, t: str) -> bool:
    pass  # TODO: implement


def _run_tests():
    assert is_isomorphic("egg", "add") is True
    assert is_isomorphic("f11", "b23") is False
    assert is_isomorphic("paper", "title") is True
    assert is_isomorphic("badc", "baba") is False
    assert is_isomorphic("a", "a") is True
    assert is_isomorphic("ab", "aa") is False
    assert is_isomorphic("foo", "bar") is False

    # --- large-input timing ---
    import time as _t

    n = 50000
    s = "ab" * (n // 2)
    t = "xy" * (n // 2)
    _t0 = _t.perf_counter()
    is_isomorphic(s, t)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f"perf is_isomorphic n={n}: {_ms:.1f}ms")
    print("all tests pass")


if __name__ == "__main__":
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/205-isomorphic-strings.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func isIsomorphic(_ s: String, _ t: String) -> Bool {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:a698c319713b4006868cfcad0c7bd59f13871a7b96204d94ebe5ad04fdff50d6
    expectEqual(Solution().isIsomorphic("egg", "add"), true, "egg-add")
    expectEqual(Solution().isIsomorphic("foo", "bar"), false, "foo-bar")
    expectEqual(Solution().isIsomorphic("paper", "title"), true, "paper-title")
    expectEqual(Solution().isIsomorphic("ab", "aa"), false, "target-reuse")
    expectEqual(Solution().isIsomorphic("a", "z"), true, "single")
    // EXCLUDED_VECTOR different-length: ["a","ab"] | The strings have equal length.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
